Properties Pane should react to selection of Editor

AlexandraBuzila opened the pull request:
- render properties of selected UI element in the editor
- highlight selection
- clear selection when UI schema changes
![editor-selection](https://user-images.githubusercontent.com/6464495/84755217-aa5d1f80-afc1-11ea-803e-720035324e00.gif)

eneufeld commented on src/editor/components/EditorElement.tsx:
```diff
@@ -66,19 +67,23 @@ export const EditorElement: React.FC<EditorElementProps> = ({
-      container
-      wrap='nowrap'
-      direction='column'
+    <div
```
why use a div?

eneufeld commented on src/editor/components/EditorElement.tsx:
```diff
@@ -66,19 +67,23 @@ export const EditorElement: React.FC<EditorElementProps> = ({
-      direction='column'
+    <div
       data-cy={`editorElement-${uiPath}`}
       className={classes.editorElement}
```
wouldn't it be easier to pass two classnames depending on the selection instead of changing the styling all the time?

AlexandraBuzila replied:
I don't think there's a difference in performance between the two approaches, but I switched to using classnames because it made the styles a bit cleaner.

eneufeld commented on src/core/util/generators/uiSchema.ts:
```diff
@@ -7,13 +7,27 @@
+  schema: SchemaElement
+): { [key: string]: any } => {
+  //TODO generate options based on schema element
+  return { multi: false };
```
I think we need a way to set a schema and uischema for the properties view, maybe it makes sense to retrieve them in the properties view instead of passing data using options

eneufeld commented on src/editor/components/EditorElement.tsx:
```diff
@@ -29,10 +29,11 @@ export interface EditorElementProps {
-  editorElement: (props: { isDragging: boolean }) => ({
-    border: '1px solid #D3D3D3',
+  editorElement: (props: { isDragging: boolean; isSelected: boolean }) => ({
+    border: props.isSelected ? '1px solid #1976d2' : '1px solid #D3D3D3',
```
maybe fill background additionally or instead of the border?

eneufeld commented on src/core/util/generators/uiSchema.ts:
```diff
@@ -7,7 +7,12 @@
+
+export const createControl = (
+  scope: string,
+  schema: SchemaElement
```
I think you forgot to remove the schema after the refactoring

sdirix commented on src/App.tsx:
```diff
@@ -51,6 +52,36 @@ const App = () => {
       .getUiSchema()
       .then((uiSchema) => dispatch(Actions.setUiSchema(uiSchema)));
   }, [schemaService]);
+  useEffect(() => {
+    //try to preserve selection when schemas change
+    const oldSelection = selection;
+    if (!oldSelection) {
+      return;
+    }
+    const newSelectedUISchemaElement =
+      oldSelection && uiSchema
+        ? getCorrespondingElement(oldSelection?.uiSchema, uiSchema)
+        : undefined;
+    const newSelectedSchemaElement =
+      oldSelection && oldSelection.schema && schema
+        ? getCorrespondingElement(oldSelection.schema, schema)
+        : undefined;
+
+    if (
+      !isPathError(newSelectedSchemaElement) &&
+      !isPathError(newSelectedUISchemaElement) &&
+      newSelectedUISchemaElement
+    ) {
+      setSelection({
+        schema: newSelectedSchemaElement,
+        uiSchema: newSelectedUISchemaElement,
+      });
+    } else {
+      //clear old selection
+      setSelection(undefined);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [uiSchema, schema]);
```
The linter rule is correct and `selection` is missing as a dependency. By not adding it as a dependency you run into the risk that it might be outdated within the effect. [See the docs](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects)
>make sure the array includes all values from the component scope (such as props and state) that change over time and that are used by the effect. Otherwise, your code will reference stale values from previous renders.

Now of course we don't want to have `selection` as a dependency as we don't want to trigger the effect when selection changes. However this is also not necessary as we actually don't need the `selection` at all. Whenever a state is updated based on its previous value, we need to use the function form of the setter, so in this case
```
setSelection(oldSelection => {});
```
This way we always get the latest state of `selection` without using old state values nor specifying it in the dependencies.

AlexandraBuzila replied:
Ah, that's a lot nicer <3

sdirix commented on src/App.tsx:
```diff
@@ -51,6 +52,36 @@ const App = () => {
+    }
+    const newSelectedUISchemaElement =
+      oldSelection && uiSchema
+        ? getCorrespondingElement(oldSelection?.uiSchema, uiSchema)
```
`getCorrespondingElement` is meant to be used only on cloned states without structural changes as it works path based. Therefore it will fail in many cases here: Whenever you select an element and insert an element somewhere "before" it in the tree, the new selection will be wrong or missing completely.

AlexandraBuzila replied:
Replaced it with `findByUUID` method

sdirix commented on src/core/selection/selection.ts:
```diff
@@ -6,5 +6,11 @@
+import { LinkedUISchemaElement } from '../model/uischema';
 
-export type SelectedElement = SchemaElement | undefined;
+export type SelectedElement = WrappedSelectedElement | undefined;
+
+interface WrappedSelectedElement {
+  uiSchema: LinkedUISchemaElement;
+  schema?: SchemaElement;
+}
```
Why do we need the `WrappedSelectedElement`, couldn't we just use `LinkedUISchemaElement`? When we need the corresponding schema element we can resolve it.

AlexandraBuzila replied:
When we set the selection, we already have the schema element. So there's no need to search for the corresponding schema element each time the selection changes. Depending on the size of the schema, that could slow things down a bit. But since we don't work with large schemas yet, I don't have a hard opinion on this.

eneufeld replied:
But the `LinkedUISchemaElement` has a list of `linkedSchemaElements`. Can't we use it?

AlexandraBuzila replied:
the `linkedSchemaElements` contains only the uuid of the schema elements, so we would have to search for the correct element in the schema.

eneufeld replied:
why can't we link the full schema? the `LinkedUISchemaElement` is not serialized is it?

AlexandraBuzila replied:
The LinkedUISchemaElements and SchemaElements are not stable, we would need to replace the instances from the `linkedSchemaElements` and `linkedUISchemaElements` on each re-render. It would also make the cloning more complex, since the schemas would not be trees anymore.

eneufeld replied:
I see, and we can't have a map to easier retrieve what we need?

sdirix commented on src/properties/components/Properties.tsx:
```diff
@@ -5,26 +5,65 @@
 export const Properties = () => {
   const [selection] = useSelection();
+  const dispatch = useDispatch();
+
```
instead of checking all places in the code whether selection exists we could do a quick exit (after all the hook calls), e.g.
```
if(!selection) {return <NoSelection />;}
```

eneufeld commented:
> Works mostly fine, however I'm not so sure about the selection handling. We should probably discuss this how we actually want to maintain the selection.

@AlexandraBuzila and I discussed this and the idea was to use UUIDs on all elements in the editor.

eneufeld commented on src/properties/propertiesService.ts:
```diff
@@ -0,0 +1,40 @@
+    uiElement: LinkedUISchemaElement,
+    schemaElement: SchemaElement | undefined
+  ): PropertySchemas | undefined => {
+    if (schemaElement && schemaElement.schema.type === 'string') {
```
format check is missing, we want to ignore the cases where we have a format defined

eneufeld commented:
I still can't check the checkbox :-(

AlexandraBuzila commented:
I added the check for format and also changed the selection to be the uuid of the selected ui element. The reason for that is that we needed to resolve the selected ui element in the UI schema anyway, when updating the properties, so there was no reason for storing the whole LinkedUIElement in the selection anymore. Now both the UI schema element and schema element need to be resolved when the selection changes. We can look at improving this approach when the need arrises (e.g use a map to quickly access the elements based on their UUID).

eneufeld commented on src/core/model/uischema.ts:
```diff
@@ -7,12 +7,14 @@
-  linkedSchemaElements?: Array<string>;
+  linkedSchemaElements?: Set<string>;
   parent?: LinkedUISchemaElement;
+  uuid?: string;
```
shoudn't this be mandatory?

AlexandraBuzila replied:
Yes, ideally it should, but if it would be mandatory we can't assign an UISchemaElement to a LinkedUISchemaElement anymore. We would need to do a cast each time we access the `elements` of a Layout, for example. But I can change this if you want.

eneufeld replied:
no keep it like it is.

sdirix replied:
For now this is fine I think.

eneufeld commented on src/core/selection/selection.ts:
```diff
@@ -5,6 +5,4 @@
-import { SchemaElement } from '../model';
-
-export type SelectedElement = SchemaElement | undefined;
+export type SelectedElement = { uuid: string } | undefined;
```
is there a reason to wrap the `uuid` in an object?

AlexandraBuzila replied:
I did this intentionally, because it more clear what the selection is supposed to be and because if we ever need to extend the SelectedElement type we can just add new properties to it without changing any code that expects the selection to contain an `uuid`

eneufeld commented on src/core/util/generators/uiSchema.ts:
```diff
@@ -5,16 +5,27 @@
-  elements: [],
-});
+export const createLayout = (type: string): LinkedUISchemaElement => {
+  const newLayout = {
```
why not directly return?

AlexandraBuzila replied:
We can return here directly, I'll fix in a new commit.

eneufeld commented on src/core/util/generators/uiSchema.ts:
```diff
@@ -5,16 +5,27 @@
+  schemaElement: SchemaElement
+): LinkedUISchemaElement => {
+  const scope = `#${getPath(schemaElement)}`;
+  const newControl = {
```
I think we can inline the scope and return directly

sdirix commented on src/App.tsx:
```diff
@@ -51,6 +52,25 @@ const App = () => {
+      }
+      return oldSelection;
+    });
+  }, [uiSchema, schema]);
```
`schema` no longer needs to be part of the dependencies
```suggestion
  }, [uiSchema]);
```

sdirix commented on src/App.tsx:
```diff
@@ -51,6 +52,25 @@ const App = () => {
+        isUUIDError(newSelectedUISchemaElement) ||
+        !newSelectedUISchemaElement
+      ) {
+        //element does not exist anymore - clear old selection
```
```suggestion
        // element does not exist anymore - clear old selection
```

sdirix commented on src/core/model/reducer.ts:
```diff
@@ -56,6 +57,7 @@ export const uiSchemaReducer = (
 ) => {
   switch (action.type) {
     case SET_UISCHEMA:
+      //FIXME we need to link the uischema to the schema
```
```suggestion
      // FIXME we need to link the uischema to the schema
```

sdirix commented on src/core/util/clone.ts:
```diff
@@ -40,23 +63,47 @@ export const getRoot = <T extends Parentable<T>>(
+    if (value && value.uuid === uuid) {
+      return value;
+    }
+    // some mappings are 'reversed'
+    if (key && key.uuid === uuid) {
+      return key;
+    }
```
These lines could be removed and it should still work as this case is also handled by the other `reversed` handling, right?
```suggestion
```

sdirix commented on src/editor/components/EditorElement.tsx:
```diff
@@ -49,36 +57,47 @@ export const EditorElement: React.FC<EditorElementProps> = ({
   const [openConfirmRemoveDialog, setOpenConfirmRemoveDialog] = React.useState(
     false
   );
+  const schemaElementUUID = wrappedElement.linkedSchemaElements?.values().next()
+    .value;
+  const searchResult = schemaElementUUID
+    ? findByUUID(schema, schemaElementUUID)
+    : undefined;
+  const elementSchema = isUUIDError(searchResult) ? undefined : searchResult;
```
Partially related to this PR: As we hardcode `linkedSchemaElements` to be of only size one, we should probably refactor it to be a simple property directly containing only one uuid. Also we can avoid resolving here but just let the reducer handle it, the reducer itself can take the uuid out of the uiSchemaElement and then do whatever it wants with it.

sdirix commented on src/editor/components/EditorElement.tsx:
```diff
@@ -49,36 +57,47 @@ export const EditorElement: React.FC<EditorElementProps> = ({
+  const classes = useEditorElementStyles();
 
+  const uiPath = getUISchemaPath(wrappedElement);
+  const isSelected = selection && selection.uuid === wrappedElement.uuid;
```
```suggestion
  const isSelected = selection?.uuid === wrappedElement.uuid;
```

sdirix commented on src/properties/components/Properties.tsx:
```diff
@@ -5,26 +5,85 @@
+  const dispatch = useDispatch();
+
+  const uiElement: LinkedUISchemaElement = useMemo(() => {
+    if (!selection || !uiSchema) return undefined;
+    const findResult = findByUUID(uiSchema, selection.uuid);
+    return isUUIDError(findResult) ? undefined : findResult;
```
I would extract this into a separate function to make the code more readable

sdirix commented on src/properties/components/Properties.tsx:
```diff
@@ -5,26 +5,85 @@
+  }, [selection, uiSchema]);
+
+  const updateProperties = useCallback(
+    ({ data: updatedProperties }) => {
+      if (!uiElement) {
+        return;
+      }
+      if (
+        !isEqual(updatedProperties, uiElement.options) &&
+        !(uiElement.options === undefined && isEmpty(updatedProperties))
+      ) {
+        dispatch(Actions.setUiSchemaOptions(uiElement, updatedProperties));
+      }
+    },
```
also extract here

sdirix commented on tsconfig.json:
```diff
@@ -13,7 +13,8 @@
     "noEmit": true,
-    "jsx": "react"
+    "jsx": "react",
+    "downlevelIteration": true
```
I would avoid adapting the `tsconfig`. When iterating over sets etc where you would need this flag you can always use `Array.from(...)` instead.
